File: netlogo-gui/src/main/prim/_other.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.agent.{Agent, AgentSet, LazyAgentSet}
import org.nlogo.api.Dump
import org.nlogo.core.I18N
import org.nlogo.nvm.{Context, Reporter}
import org.nlogo.nvm.RuntimePrimitiveException
import java.util.ArrayList

class _other extends Reporter {

  def report(context: Context): AgentSet = {
    report_1(context, argEvalAgentSet(context, 0))
  }

  def report_1(context: Context, sourceSet: AgentSet): AgentSet = {
    if (sourceSet.isInstanceOf[LazyAgentSet]) {
      sourceSet.asInstanceOf[LazyAgentSet].lazyOther(context.agent)
      sourceSet
    } else {
//      new LazyAgentSet(null, sourceSet, context.agent :: Nil)
      val others = new ArrayList[Agent]
      others.add(context.agent)
      new LazyAgentSet(null, sourceSet, others)
    }
  }
}

File: netlogo-gui/src/main/prim/_other.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.agent.Agent;
import org.nlogo.agent.AgentIterator;
import org.nlogo.agent.AgentSet;
import org.nlogo.agent.AgentSetBuilder;
import org.nlogo.api.LogoException;
import org.nlogo.core.Syntax;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.Reporter;

public final class _other
    extends Reporter {
  // super( "TPL" ) ;


  @Override
  public Object report(final Context context)
      throws LogoException {
    return report_1(context, argEvalAgentSet(context, 0));
  }

  public AgentSet report_1(final Context context, AgentSet sourceSet) {
    AgentSetBuilder result = new AgentSetBuilder(sourceSet.kind(), sourceSet.count());
    for (AgentIterator it = sourceSet.iterator(); it.hasNext();) {
      Agent otherAgent = it.next();
      if (context.agent != otherAgent) {
        result.add(otherAgent);
      }
    }
    return result.build();
  }
}
